servers/src/main/java/tachyon/master/block/BlockMaster.java

```diff
@@ -561,6 +573,16 @@ private void processWorkerRemovedBlocks(MasterWorkerInfo workerInfo,
   }
 
   /**
+   * Called by the heartbeat thread whenever a worker is lost.
+   * @param latest the latest {@link MasterWorkerInfo} available at the time of worker loss
+   */
+  // Synchronized on mBlocks by the caller
+  private void processLostWorker(MasterWorkerInfo latest) {
```

Comment:
Thanks for pointing the deadlock issue out, @gpang. I fixed that in the most simple way I could think of, i.e., locking all the `heartbeat()` section with `mBlocks`. I could think of other solutions (for example, queuing the processing of the lost worker in a queue synchronized separately), but they would all to be far too messy imho. Moreover, the execution of `heartbeat()` should be relatively brief anyway.